### PR TITLE
Module3 project exercise

### PR DESCRIPTION
**Pull request by LouisLon**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,25 +1,42 @@
 # DevOps Apprenticeship: Project Exercise
```
Comment (Jonesey13) [inline README.md:1]: In this file you're missing:

- Notes on setting up the Trello Config
- Instructions on how to run the tests

File: todo_app/.env.test [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# Flask server configuration.
+FLASK_APP=todo_app/app
+FLASK_ENV=Testing
+FLASK_SKIP_DOTENV=True
+
+# Change the following values for local development.
+SECRET_KEY=secret-key
+TRELLO_KEY=sadasdasdadqweqweqwe45345
```
Comment (Jonesey13) [inline todo_app/.env.test:8]: I hope these are not valid trello credentials :p

File: todo_app/app.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    @app.route('/')
+    def index():  
+        try:        
+            #return "hello"              
+            items = session.get_items()
+            item_view_model = session.ViewModel(items)
+            return render_template('index.html',view_model=item_view_model)
+            
+        except Exception as e:
+            return str(e)
```
Comment (Jonesey13) [inline todo_app/app.py]: This removes the stack trace making this harder to debug

File: todo_app/data/session_items.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
+from flask import session
```
Comment (Jonesey13) [inline todo_app/data/session_items.py]: You must not rely on using the flask session cookie. This currently breaks your integration test.

This is a blocker for this passing review.

File: todo_app/tests/app_test.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    requests_mock.get('https://api.trello.com/1/search', text=board_data)
+    requests_mock.get('https://api.trello.com/1/boards/5f58c9c553d5e50e89e24311/lists', text=listinboard_data)
+    requests_mock.get('https://api.trello.com/1/boards/5f58c9c553d5e50e89e24311/cards', text=cardsinboard_data)
+    
+    assert 'Book my holiday' in client.get('/').data.decode("utf-8")
```
Comment (Jonesey13) [inline todo_app/tests/app_test.py:90]: This test doesn't work as the flask session is not active during the lifecycle of this test

File: todo_app/tests/app_test.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    requests_mock.get('https://api.trello.com/1/search', text=board_data)
+    requests_mock.get('https://api.trello.com/1/boards/5f58c9c553d5e50e89e24311/lists', text=listinboard_data)
+    requests_mock.get('https://api.trello.com/1/boards/5f58c9c553d5e50e89e24311/cards', text=cardsinboard_data)
+    
+    assert 'Book my holiday' in client.get('/').data.decode("utf-8") 
+
+# def test_cant_add_number_to_string():
```
Comment (Jonesey13) [inline todo_app/tests/app_test.py]: All this commented out code can go (you won't lose it as it's in your source control history)

File: todo_app/tests/flask_config.py [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import os
```
Comment (Jonesey13) [inline todo_app/tests/flask_config.py]: When the flask session cookie is removed this file can go

**Review by Jonesey13 — CHANGES_REQUESTED** [review]
Nearly there. The integration test does not work due to the flask session usage.

Also these changes are not mergeable atm. Your changes should be on top of the Module 2 branch (or the latest master). Consider merging master in and resolving the conflicts.

File: todo_app/data/session_items.py (hunk 4 above)
Comment (LouisLon) [reply todo_app/data/session_items.py]: Could you give more clarity on what you mean by not using flask session cookie

File: todo_app/app.py (hunk 3 above)
Comment (LouisLon) [reply todo_app/app.py]: Could you point me in the right direction as I don't understand what you mean here

File: todo_app/.env.test (hunk 2 above)
Comment (LouisLon) [reply todo_app/.env.test:8]: no this is not a valid credential, should I leave it blank

File: todo_app/app.py (hunk 3 above)
Comment (LouisLon) [reply todo_app/app.py]: please I need to understand how this removes stack trace

File: todo_app/tests_e2e/app_e2e_test.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+from selenium.webdriver.common.keys import Keys
+from selenium.webdriver.common.by import By
+
+@pytest.fixture(scope='module')
+def test_app():
+# Create the new board & update the board id environment variable    
+    file_path = find_dotenv('.env')
+    load_dotenv(file_path, override=True)
+    board_id = session.create_trello_board()
+    os.environ['TRELLO_BOARD_ID'] = board_id
```
Comment (hugh-emerson) [inline todo_app/tests_e2e/app_e2e_test.py:17]: What do you imagine this line is doing?

Your app doesn't read TRELLO_BOARD_ID anywhere. It finds by board name. If you want to keep that, then you need to make the board name configurable.

It also looks for list by name, and doesn't use the default list names. So you either need to change that/make it configurable or set up the test board with the correct list names.

Currently your selenium test uses the real board.

File: todo_app/tests_e2e/app_e2e_test.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    # construct the new application
+    application = app.create_app()
+    # start the app in its own thread.
+    thread = Thread(target=lambda: application.run(use_reloader=False))
+    thread.daemon = True
+    thread.start()
+    yield app
+    # Tear Down
+    thread.join(1)
+
```
Comment (hugh-emerson) [inline todo_app/tests_e2e/app_e2e_test.py:43]: What's the point of this fixture? Delete it and just use `test_app`

If you were worried about recreating the board, `@pytest.fixture(scope='module')` means it won't have to re-run within the module.

**Review by hugh-emerson — COMMENTED** [review]
Your selenium test doesn't use the temporary test board. See my other comment for details.
Also, you should add geckodriver.log to the .gitignore

But otherwise things seem to work. 👍

**Pull request merged**